docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,240 @@
+
+A pre-event handler is declared using a `pre event handler` declaration:
+
+    pre event handler for <name:id> is <handler:constexpr>
+
+By default, all pre-event handlers declared by the list of interested controls are run.  There is no need to "pass" the event onwards [1].
+
+A control can use the pre-event handler to **capture** an event:
+
+    capture event
+
+The `capture event` statement makes the capturing control becomes the "target" for the event; all its children are removed from the list of interested controls [2,3].
+
+**Notes:**
+
+1. As per the `mouseEnter`/`mouseLeave`, implicit passing reflects the expectation that pre- and post-event handlers are usually be used to implement behaviour that complements rather than conflicts with that of other interested controls.
+
+2. This allows composed widgets to "hide" events from their children.
+
+3. Since `capture event` causes the capturing control to become the target, the capturing control can't do anything that requires knowledge of the "original" target in its on-event handler unless it saves `the event target` in the pre-event handler before `capture event`.  On the other hand, there's not much point in having separate pre-event and on-event handlers when capturing because they're guaranteed to be run immediately one after another.
+
+### Dispatch phase
+
+During the dispatch phase, **on-event** handlers are run.  They are run from "inside-out", starting from the target control.  A parent widget's on-event handler runs after its child widgets' on-event handlers.
+
+An on-event handler is declared using an `event handler` declaration [1]:
+
+    [on] event handler for <name:id> is <handler:constexpr>
+
+By default, if an on-event handler is found for an event, then it **blocks** further on-event handler execution for the event [2].
+
+An on-event handler can use the `pass event` statement to prevent the default-blocking behaviour.
+
+The `pass event` statement does not return.
+
+If an on-event handler yields, it is treated as having blocked further on-event handler execution.  It is an error to `pass event` after yielding [3].
+
+**Notes:**
+
+1. Since most widgets will only need to handle events during the dispatch phase, on-event handlers may be declared without a qualifier.
+
+2. This is for conceptual compatibility with LCS.
+
+3. Since yielding detaches an event handler from event processing, there is subsequently no "next handler" for the on-event handler to pass the event on to.
+
+### Leave phase
+
+During the leave phase, **post-event** handlers are run.  Like on-event handlers, they are run from "inside-out", starting from the target control.
+
+All interested controls have the opportunity to run a post-event handler, even if an on-event handler blocked the event during the dispatch phase [1].
+
+A post-event handler is declared using a `post event handler` declaration:
+
+    post event handler for <name:id> is <handler:constexpr>
+
```

Comment:
@peter-b Regarding (1) - I know that and you know that, but you as you state in (2) lots of LiveCode users don't do that. (2) should definitely be discouraged but it also involves a deeper understanding of why code should be written that way. 

Coding becomes more complicated when you start thinking about async operations, disabling controls while an operation is performed, or executing a handler AFTER the mouseUp event has completed using <send>. While these types of approaches are what I use in my apps, I don't think LiveCode has a LiveCode-like approach to dealing with these problems. For example, <send MESSAGE in 0 seconds> isn't self-documenting at all.

But the original question is about when the post-event handler in the widget should be triggered. Without taking any of the implementation details into account, I would say the messages between LCS and LCB should be balanced. If that proves too difficult given the types of things that can go on in LCS during any given event then the world won't end.